ql/src/java/org/apache/hadoop/hive/ql/exec/tez/HostAffinitySplitLocationProvider.java

```diff
@@ -72,11 +78,17 @@ public HostAffinitySplitLocationProvider(List<String> knownLocations) {
     FileSplit fsplit = (FileSplit) split;
     String splitDesc = "Split at " + fsplit.getPath() + " with offset= " + fsplit.getStart()
         + ", length=" + fsplit.getLength();
-    List<String> preferredLocations = preferLocations(fsplit);
-    String location =
-        preferredLocations.get(determineLocation(preferredLocations, fsplit.getPath().toString(),
-            fsplit.getStart(), splitDesc));
-    return (location != null) ? new String[] { location } : null;
+    List<String> preferredLocations = new ArrayList<>(preferLocations(fsplit));
+    List<String> finalLocations = new ArrayList<>(numberOfLocations);
+    // Generate new preferred locations until we need more, or we do not have any preferred
+    // location left
+    while (finalLocations.size() < numberOfLocations && preferredLocations.size() > 0) {
+      String nextLocation = preferredLocations.get(determineLocation(preferredLocations,
+          fsplit.getPath().toString(), fsplit.getStart(), splitDesc));
+      finalLocations.add(nextLocation);
+      preferredLocations.remove(nextLocation);
```

Comment:
Instead of having an own array list, from which we remove 1 for each split (which is quite often), we could simply pass down numberOfLocations (and decrement it) to determineLocation and change

`int index = Hashing.consistentHash(hash1, locations.size());`
to
`int index = Hashing.consistentHash(hash1, numLocations);`

in determineLocations. Seems to be more efficient.

Reply:
Something else to consider: let's say that we generate multiple locations for each split now, we will run through the determineLocation multiple times (vs. once) to generate locations that most likely are not used later down the code path. This means that we spend more cycles on TezAM, potentially increasing the query execution time there always where the benefits are only seen in rejection cases....

Reply:
I did some measurements for the split generation with this code:
`
@Test (timeout = 5000000)
public void testOrcSplitsBasic() throws IOException {
  HostAffinitySplitLocationProvider locationProvider = new HostAffinitySplitLocationProvider(executorLocations, true, 1);

  InputSplit os1 = createMockFileSplit(true, "path1", 0, 1000, new String[] {locations.get(0), locations.get(1), locations.get(2), locations.get(3)});

  long start = System.nanoTime();
  for(int i=0;i<100000;i++) {
    locationProvider.getLocations(os1);
  }
  LOG.error("TIME: " + (System.nanoTime()-start)/1000000);
}
`

I got the following results:
Original code (~6100ms for 100k requests):
- 5859
- 6511
- 6813
- 5721
- 5663

New code with 1 location (~5823ms for 100k requests):
- 5877
- 5621
- 5613
- 5883
- 6120

New code with 2 locations (~6579ms for 100k request):
- 6433
- 6825
- 6574
- 6444
- 6621

I do not see why the new code should be faster, so this means probably high variation for the data. Generating 2 locations instead of 1 seems like a 10% overhead. Since this is 0.006ms per request this seems reasonable for me.

What is your opinion?